core/blockchain/production/impl/block_producer_impl.cpp

```diff
@@ -0,0 +1,151 @@
+    OUTCOME_TRY(parent_weight,
+                chain_weight_calculator_->calculateWeight(parent_tipset));
+    std::vector<SignedMessage> messages =
+        message_storage_->getTopScored(config::kBlockMaxMessagesCount);
+    OUTCOME_TRY(msg_meta, getMessagesMeta(messages));
+    std::vector<UnsignedMessage> bls_messages;
+    std::vector<SignedMessage> secp_messages;
+    std::vector<crypto::bls::Signature> bls_signatures;
+    for (auto &&message : messages) {
+      visit_in_place(
+          message.signature,
+          [&message, &bls_messages, &bls_signatures](
+              const BlsSignature &signature) {
+            bls_messages.emplace_back(std::move(message.message));
+            bls_signatures.push_back(signature);
+          },
+          [&message, &secp_messages](const Secp256k1Signature &signature) {
+            secp_messages.emplace_back(std::move(message));
+          });
+    }
+    OUTCOME_TRY(bls_aggregate_sign,
+                bls_provider_->aggregateSignatures(bls_signatures));
+    Time now = clock_->nowUTC();
+    OUTCOME_TRY(current_epoch, epoch_->epochAtTime(now));
+    BlockHeader header{
+        .miner = std::move(miner_address),
+        .ticket = ticket,
+        .epost_proof = std::move(proof),
+        .parents = std::move(parent_tipset.cids),
+        .parent_weight = parent_weight,
+        .height = static_cast<uint64_t>(current_epoch),
+        .parent_state_root = std::move(vm_result.state_root),
+        .parent_message_receipts = std::move(vm_result.message_receipts),
+        .messages = msg_meta.getCID(),
+        .bls_aggregate = {bls_aggregate_sign.begin(), bls_aggregate_sign.end()},
+        .timestamp = static_cast<uint64_t>(now.unixTime().count()),
+        .block_sig = {},  // Block must be signed be Actor Miner
+        .fork_signaling = 0};
+    Block block{.header = std::move(header),
+                .bls_messages = std::move(bls_messages),
+                .secp_messages = std::move(secp_messages)};
+    return std::move(block);
+  }
+
+  outcome::result<BlockProducerImpl::Tipset> BlockProducerImpl::getTipset(
+      const CID &tipset_id) const {
+    auto raw_data = data_storage_->get(tipset_id);
+    if (raw_data.has_error()) {
+      return BlockProducerError::PARENT_TIPSET_NOT_FOUND;
+    }
+    gsl::span<const uint8_t> raw_span{raw_data.value().toVector()};
+    auto tipset = codec::cbor::decode<Tipset>(raw_span);
+    if (tipset.has_error()) {
+      return BlockProducerError::PARENT_TIPSET_INVALID_CONTENT;
+    }
+    return tipset.value();
+  }
+
+  outcome::result<BlockProducerImpl::MsgMeta>
+  BlockProducerImpl::getMessagesMeta(
+      const std::vector<SignedMessage> &messages) {
+    auto bls_backend = std::make_shared<InMemoryDatastore>();
+    auto secp_backend = std::make_shared<InMemoryDatastore>();
+    Amt bls_messages_amt{bls_backend};
+    Amt secp_messages_amt{secp_backend};
+    for (size_t index = 0; index < messages.size(); ++index) {
```

Comment:
Split messages by type first, storing cids to separate arrays.
Then build two separate Amt/adt::Array from corresponding cid arrays.